snuba/consumers/consumer.py

```diff
@@ -616,40 +639,70 @@ def process_message_multistorage(
             if (index < len(message.payload.storage_keys) - 1)
             else value
         )
-        result = (
-            get_writable_storage(storage_key)
-            .get_table_writer()
-            .get_stream_loader()
-            .get_processor()
-            .process_message(storage_message, metadata)
+
+        result = process_message_multistorage_work(
+            metadata=metadata, storage_key=storage_key, storage_message=storage_message
         )
-        if isinstance(result, AggregateInsertBatch):
-            values_row_encoder = get_values_row_encoder(storage_key)
-            results.append(
-                (
-                    storage_key,
-                    BytesInsertBatch(
-                        [values_row_encoder.encode(row) for row in result.rows],
-                        result.origin_timestamp,
-                    ),
-                )
-            )
-        elif isinstance(result, InsertBatch):
-            results.append(
-                (
-                    storage_key,
-                    BytesInsertBatch(
-                        [json_row_encoder.encode(row) for row in result.rows],
-                        result.origin_timestamp,
-                    ),
-                )
-            )
-        else:
-            results.append((storage_key, result))
+        results.append((storage_key, result))
 
     return results
 
 
+def process_message_multistorage_identical_storages(
+    message: Message[MultistorageKafkaPayload],
+) -> Sequence[Tuple[StorageKey, Union[None, BytesInsertBatch, ReplacementBatch]]]:
+    """
+    This method is similar to process_message_multistorage except for a minor difference.
+    It performs an optimization where it avoids processing a message multiple times if the
+    it finds that the storages on which data needs to be written are identical. This is a
+    performance optimization since we remove the message processing time completely for all
+    identical storages like errors and errors_v2.
+
+    It is possible that the storage keys in the message could be a mix of identical and
+    non-identical storages. This method takes into account that scenario as well.
+
+    The reason why this method has been created rather than modifying the existing
+    process_message_multistorage is to avoid doing a check for every message in cases
+    where there are no identical storages like metrics.
```

Comment:
Are you sure about doing this optimization? It seems the check is extremely cheap.

Reply:
Yes I want to keep this optimization for now. Don't want to impact other storages when only a few would be affected.